### PR TITLE
feat(cdn): can have multiple buffer

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Review by richardlt — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

File: engine/cdn/storage/storage_unit_io.go [hunk 1]
```diff
@@ -105,9 +106,12 @@ func (r RunningStorageUnits) NewSource(ctx context.Context, refItemUnit sdk.CDNI
 	}
 	var unit source = r.Storage(refUnit.Name)
 	if unit == nil {
-		if r.Buffer.Name() == refUnit.Name {
-			unit = r.Buffer
-		} else {
+		for _, bu := range r.Buffers {
+			if bu.Name() == refUnit.Name {
+				unit = bu
```
Comment (yesnault) [inline engine/cdn/storage/storage_unit_io.go:111]: break

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/cdn/cdn_item.go [hunk 2]
```diff
@@ -310,14 +310,15 @@ func (s *Service) completeItem(ctx context.Context, tx gorpmapper.SqlExecutorWit
...
 	var reader io.ReadCloser
-	switch itemUnit.UnitID {
-	case s.Units.Buffer.ID():
-		// Get all data from buffer and add manually last line
-		reader, err = s.Units.Buffer.NewReader(ctx, itemUnit)
-		if err != nil {
-			return err
+	for _, unit := range s.Units.Buffers {
+		if unit.ID() == itemUnit.UnitID {
+			reader, err = unit.NewReader(ctx, itemUnit)
```
Comment (yesnault) [inline engine/cdn/cdn_item.go:315]: break?

**Review by yesnault — CHANGES_REQUESTED** [review]

File: engine/cdn/cdn_sync.go [hunk 3]
```diff
@@ -38,35 +38,39 @@ func (s *Service) startCDSSync(ctx context.Context) error {
...
+			}
+			if sdk.ErrorIs(err, sdk.ErrNotFound) {
+				if err := bu.Remove(ctx, sdk.CDNItemUnit{ItemID: itemID}); err != nil {
+					log.Error(ctx, "[SyncBuffer] unable to remove item %s from buffer: %v", itemID, err)
+					continue
+				}
+				keysDeleted++
+				log.Info(ctx, "[SyncBuffer] item %s remove from redis", itemID)
+			} else {
+				log.Error(ctx, "[SyncBuffer] unable to load item %s", itemID)
```
Comment (yesnault) [inline engine/cdn/cdn_sync.go:69]: missing err in log

**Review by yesnault — CHANGES_REQUESTED** [review]

**Review by yesnault — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15203.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Pull request merged**